---
title: "R more"
output: 
    html_document:
        theme: cerulean
        css: style.css
---

# Presentations

* [Introduction](slides/introduction.html) ([.Rmd](slides/introduction.Rmd))

# Topics

* [Rmarkdown](topics/Rmarkdown.html) ([.Rmd](topics/Rmarkdown.Rmd))
* [Programming](topics/programming.html) ([.Rmd](topics/programming.Rmd))
* The tidyverse: tidyr, dplyr, ggplot2
* [Sequences and genomic features](topics/sequences_and_features.html) ([.Rmd](topics/sequences_and_features.Rmd))
* Shiny

# See also

Also in this series:

* [R Introduction](https://monashbioinformaticsplatform.github.io/r-intro/)
* [RNAseq using R](http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/)

Other resources:

* [Next steps](https://monashbioinformaticsplatform.github.io/r-intro/next.html)


# Authors

This course is developed by the Monash Bioinformatics Platform:

* Paul Harrison
* Adele Baraguhare
* Kirill Tsyganov

<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

Some of this material is derived from work by [Software Carpentry](http://software-carpentry.org/). Any errors and failures of pedagogy are our own.

# Source code

* [GitHub repository](https://github.com/MonashBioinformaticsPlatform/r-more)




