---
title: "Programming"
output: html_document
---

In the introductory R day, we mostly used R interactively, but one of the final topics introduced our first programming concept, [for-loops](https://monashbioinformaticsplatform.github.io/r-intro/for_loops.html). Today we will expand our knowledge of programming by writing functions and if-statements, and apply this knowledge to automate a typical task from bioinformatics.

We are using a couple of packages from CRAN in this tutorial, which we can install with `install.packages`:

```{r eval=FALSE}
# Don't run this if you are using our biotraining server, the packages are already installed!
install.packages(c(
    "readr",    # read tabular data
    "dplyr"     # general data frame manipulation
))
```


# Functions

In the introductory R day, we used a lot of functions that other people had written. We're now going to see how to create our own functions.

```{r}
fahr_to_kelvin <- function(temp) {
    (temp-32) * (5/9) + 273.15
}
```

We define `fahr_to_kelvin` by assigning it a `function`. The list of argument names are containted within parentheses. Next, the body of the function--the statements that are executed when it runs--is contained within curly braces (`{}`). The statements in the body have been indented by four spaces, which makes the code easier to read but does not affect how the code operates.

When we call the function, the values we pass to it are assigned to those variables so that we can use them inside the function.
The expression inside the function is evaluated and the result is "returned" back to whoever asked for it.

Let's try running our function. Calling our own function is no different from calling any other function:

```{r}
# freezing point of water
fahr_to_kelvin(32)
# boiling point of water
fahr_to_kelvin(212)
```


## Variations

Here are some other equivalent ways to write this function. 

The body of the function within the `{}` can contain multiple lines of code, "statements". Only the last line is returned.

```{r}
fahr_to_kelvin <- function(temp) {
    kelvin <- (temp-32) * (5/9) + 273.15
    kelvin
}
```

The braces are not actually needed if the function only contains one statement.

```{r}
fahr_to_kelvin <- function(temp) (temp-32) * (5/9) + 273.15
```

`return` can be used to return from a function immediately. Further code will not be run.

```{r}
fahr_to_kelvin <- function(temp) {
    kelvin <- (temp-32) * (5/9) + 273.15
    return(kelvin)
    plot(1:10)
}
```

Statements in the function may be split over several lines, so long as it is clear that the statement is incomplete due to an unclosed bracket or an operator in need of a right hand argument.

```{r}
fahr_to_kelvin <- function(temp) {
    kelvin <- 
        (temp-32) * (5/9) + 
        273.15
    return(kelvin)
}
```

Spacing and layout is largely for our own sanity.

```{r}
   fahr_to_kelvin<- 
function(
                   temp){(temp 
  -32  )*(    5 
            /9 )+ 

273.15}
```

Beware! however of accidentally finishing a statement early.

```{r}
fahr_to_kelvin_broken <- function(temp) {
    (temp-32) * (5/9) 
        + 273.15
}

fahr_to_kelvin_broken(212)
```

The first line in the body of the function is computed then discarded. The second line, `+ 273.15`, is valid R so we don't even get an error.

Finally, the computer doesn't understand or care what names we give to functions, arguments, or variables.

```{r}
charmander <- function(bulbasaur) {
    mew <- (bulbasaur-32) * (5/9) + 273.15
    mew
}
```


## Composing Functions

Now that we've seen how to turn Fahrenheit into Kelvin, it's easy to turn Kelvin into Celsius:

```{r}
kelvin_to_celsius <- function(temp) {
    temp - 273.15
}

#absolute zero in Celsius
kelvin_to_celsius(0)
```

What about converting Fahrenheit to Celsius? We could write out the formula, but we don't need to. Instead, we can compose the two functions we have already created:

```{r}
fahr_to_celsius <- function(temp) {
    temp_k <- fahr_to_kelvin(temp)
    temp_c <- kelvin_to_celsius(temp_k)
    temp_c
}

# freezing point of water in Celsius
fahr_to_celsius(32.0)
```

This is our first taste of how larger programs are built: we define basic operations, then combine them in ever-larger chunks to get the effect we want. Real-life functions will usually be larger than the ones shown here--typically half a dozen to a few dozen lines--but they shouldn't ever be much longer than that, or the next person who reads it won't be able to understand what's going on.

You might have noticed that all of these functions have an argument called `temp`. Why hasn't this caused confusion and chaos? The answer is that arguments to a function and variables defined in the body of a function are *local* to a call to that function. Within the body of a function, R code can see variables from its own local environment, and variables from the global environment, but *not* variables local to other function calls. While running code R maintains a *stack* of local environments. When a call to a function returns, all of its local arguments and variables disappear.

<b>Tip:</b> There are several ways to see this in action.

* Scatter `cat`s through your functions showing the values of arguments and variables at various points. 

* Insert a call to `browser()` in one of your functions. This will pause execution and let you interact with the local environment in a function. Once in the browser, type "help" for a list of commands. Besides examining variables, you can step through the code from the point it paused with "s" or "n". When you are done type "f".

* You can debug an existing function using `debugonce`. This is like temporarily adding a call to `browser()` at the top of the function.

```{r, eval=FALSE}
debugonce(fahr_to_celsius)
fahr_to_celsius(212)
```


## Challenge {.challenge}

Write a function to calculate the length of the hypotenuse of a right angled triangle using Pythagorus's rule, given the lengths of the other sides.

Hint: `sqrt` calculates the square root of a number.

Testing your code is important. Invent a test case for your code consisting of:

* The input arguments to your function.
* The return value you expect.

Confirm that your function works as expected.


# A practical programming exercise

Let's say we want to examine the quality of some FASTQ files, which contain reads from a DNA sequencing machine. An experiment has been performed over several days, and we want to run a program called "fastqc" on each of the FASTQ files.

FASTQ is a simple text format containg a series of DNA sequences and associated quality information. Examine "Day0.fastq" in the "r-more-files" folder.

(This data is a very small portion of the reads from an experiment inducing pluripotency in a mouse cell line, [GSE70022](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70022).)

## Running external software

Software outside of R is normally run from a "terminal" window, which is rather like the console in RStudio. You type in a command and the operating system runs it for you. The commands you type in are interpreted by something called a "shell", because it acts like a shell around the operating system. For more information, see the [Software Carpentry](http://software-carpentry.org/lessons/) course on the Unix shell.

R can give a command to the shell using the `system` function. For example the "uptime" command tells you how long the computer has been running for.

```{r}
system("uptime")
```

Here is how we want to run FastQC for day 0. To R the command is just a character string. It will be interpreted by the shell, which is external to R. The shell will then run FastQC for us.

```{r cache=TRUE}
system("fastqc --extract --outdir . r-more-files/Day0.fastq")
```

In the "Files" tab of the bottom right pane in RStudio, you should see that some new files and a directory have been created. Click on "Day0_fastqc.html" to examine it. Also examine the file in the "Day0_fastqc" folder called "summary.txt".

<b>If you don't have FastQC, or it failed to run for some reason:</b> The expected output files can be found in the folder "r-more-files/fastqc-output".

## For-loops revisited

Recall in the introductory R day we used for-loops to perform repetitive tasks.

```{r results="hold"}
for(i in c("a","b","c","d","e")) {  #    1
    cat("i is",i,"\n")              #      2   4   6   8   10
    cat("potato\n")                 #        3   5   7   9    11
}                                   #
cat("done\n")                       #                            12
                                    #   --order-of-execution-->
                                    #
                                    # i= - a a b b c c d d e  e  e ...
```

Typing out the fastqc command for each day will get repetitive. Let's use a for-loop to automate this task. The `paste0` function lets us "paste" together character strings, so we can use that to construct the commands to run, like this:

```{r}
day <- 0
command <- paste0("fastqc --extract --outdir . r-more-files/Day", day, ".fastq")
command
```

We want to do this for each day, and then use `system` to run the resulting command:

```{r cache=TRUE}
days <- c(0,4,7,10,15,20)

for(day in days) {
    command <- paste0("fastqc --extract --outdir . r-more-files/Day", day, ".fastq")
    cat("Running the command:", command, "\n")
    system(command)
}
```


## Loading the summary.txt files

**Note:** If you weren't able to run FastQC earlier, the expected output files can be found in the folder "r-more-files/fastqc-output". You will need to adjust the filenames appropriately in the R code below.

The summary.txt files are in "tab separated value" format. This is similar to comma separated value format we've seen in the introductory R day, but instead of using a comma it uses a special character called a tab to delimit columns. Tabs show up as variable amounts of space in a text editor. In R, they can be written `"\t"`.

In base R we could use the function `read.delim` to read this file. It's quite similar to `read.csv`. 

However, let's use one of Hadley Wickham's packages, `readr`, to load the file. If this package isn't installed, you can install it with:

```{r eval=FALSE}
install.packages("readr")
```

To use `readr` we first need to load it with `library`.

```{r}
library(readr)
```

We can now read the file like this:

```{r}
read_tsv("Day0_fastqc/summary.txt")
```

Oh! There aren't column headings in this file. We need to tell `read_tsv` this.

```{r}
read_tsv("Day0_fastqc/summary.txt", col_names=FALSE)
```

We should tidy this up a bit before using it. Columns should have meaningful names, and the PASS/WARN/FAIL grading looks like it should be a factor:

```{r}
filename <- "Day0_fastqc/summary.txt"
sumtab <- read_tsv(filename, col_names=FALSE)
colnames(sumtab) <- c("grade", "test", "file")
sumtab$grade <- factor(sumtab$grade, c("FAIL","WARN","PASS"))
sumtab
```

We expect to have to examine many FastQC reports in future. It will be convenient to have this as a function!

`filename` is going to be different each time we use this code, so it wants to be an argument to the function. The rest of the code goes in the body of the function. The returned value will be `sumtab`.

```{r}
load_fastqc <- function(filename) {
    sumtab <- read_tsv(filename, col_names=FALSE)
    colnames(sumtab) <- c("grade", "test", "file")
    sumtab$grade <- factor(sumtab$grade, c("FAIL","WARN","PASS"))
    sumtab
}

load_fastqc("Day0_fastqc/summary.txt")
```


## Applying the function

We now want to load each of the `summary.txt` files we created earlier. One perfectly valid way would be to use a for-loop. An example of this was given in the introductory R day. However here we're going to use an apply-style function.

In the introductory R day we saw a couple of flavours of `apply`. `apply` itself was for use with matrices. `tapply` was used with data frames and could produce a vector or a matrix or even a higher dimensional array. Another flavour is apply is `lapply` which produces lists. As lists can hold anything `lapply` can be used with any function, no matter what type of thing the function returns.

First let's work out the names of the files we want to load.

```{r}
days <- c(0,4,7,10,15,20)
filenames <- paste0("Day", days, "_fastqc/summary.txt")
filenames
```

Now we can use `lapply` to apply our function to each of the filenames.

```{r}
sumtabs <- lapply(filenames, load_fastqc)
```

<b>Tip:</b> If you wanted to process multiple vectors in parallel you could use the `Map` function or Hadley Wickham's `purrr` package.

We now have a list of data frames. We want to bind the data frames together into one big data frame. One way would be to use `rbind` and do it step by step in a for loop. Here are two other ways:

```{r message=F, warning=F}
# Base R way
bigtab <- do.call(rbind, sumtabs)

# The dplyr way
library(dplyr)
bigtab <- bind_rows(sumtabs)

bigtab
table(bigtab$test, bigtab$grade)
```

(If `dplyr` isn't installed, you would need to install it with `install.packages`.)

<b>Take home message:</b> All of the apply-style functions are enormously more powerful now that we can supply them with functions we write ourselves.



# If statements

One final programming concept you need to know about is the "if"-statement. An if-statement lets us do something only if some logical value is `TRUE`, or do one thing if it is `TRUE` and another if it is `FALSE`.

In the introductory R day, we used logical vectors to query a data frame. Everything we learned about making and manipulating logical vectors is applicable here, but with the restriction that we need a single logical value, i.e. a logical vector of length 1.

```{r, results="hold"}
num <- 37                   # 1
if (num > 100) {            #   2
  cat("greater\n")          #
} else {                    #
  cat("not greater\n")      #     3
}                           #       
cat("done\n")               #       4
                            # --time-->
```

The second line of this code uses an if-statement to tell R that we want to make a choice. If the following test is `TRUE`, the body of the `if` (i.e., the lines in the curly braces underneath it) are executed. If the test is `FALSE`, the body of the `else` is executed instead.
Only one or the other is ever executed.

In the example above, the test `num > 100` returns the value `FALSE`, which is why the code inside the `if` block was skipped and the code inside the `else` block was run instead.

If-statements don't have to include an `else`. If there isn't one, R simply does nothing if the test is FALSE:

```{r results="hold"}
num <- 53                            # 1
if (num > 100) {                     #   2
  cat("num is greater than 100\n")   #
}                                    #     
cat("done\n")                        #     3
                                     # --time-->
```


We can also chain several tests together when there are more than two options. As an example, here is a function that returns the sign of a number:

```{r}
sign <- function(num) {
    if (num > 0) {            # line 1
        return(1)             # line 2
    } else if (num == 0) {    # line 3
        return(0)             # line 4
    } else {                  
        return(-1)            # line 5
    }                         
}

sign(-3)
sign(0)
sign(2/3)
```

## Quiz {.challenge}

Which lines of the function `sign` executed when it was called above, and in what order?


## Improving load_fastqc

Our `load_fastqc` function will currently fail with an error if the file it is passed does not exist.

```{r error=TRUE}
load_fastqc("nosuchfile.txt")
```

Depending on how we intend to use it, we might instead want the function to issue a warning and return NULL.

```{r}
load_fastqc <- function(filename) {
    # Check arguments are sane
    if (!file.exists(filename)) {
        warning("No such file: ", filename)
        return(NULL)
    }
    
    # Load and tidy data
    sumtab <- read_tsv(filename, col_names=FALSE)
    colnames(sumtab) <- c("grade", "test", "file")
    sumtab$grade <- factor(sumtab$grade, c("FAIL","WARN","PASS"))
    sumtab
}

load_fastqc("nosuchfile.txt")
```

Checking input arguments before proceeding is a common pattern when writing R functions.


# Sourcing .R files

Having developed some useful functions, we might want to re-use them in a future project or share them with a friend. We can store R code in a file, usually with extension ".R", and load it with the `source` function.

Put your `load_fastqc` function in a file called "fastqc.R". It uses the `readr` library, so be sure to load the library in the file as well.

```
# fastqc.R file should contain:

library(readr)

load_fastqc <- function(filename) {
    sumtab <- read_tsv(filename, col_names=FALSE)
    colnames(sumtab) <- c("grade", "test", "file")
    sumtab$grade <- factor(sumtab$grade, c("FAIL","WARN","PASS"))
    sumtab
}
```

```{r eval=FALSE}
# From the console:

source("fastqc.R")
```

Everything in the file runs as though it were typed on the console.

Similarly, if you are going to be doing several different things with a data set, you could write a .R file to load the data into a tidy form, and several other .R scripts to do various different things with it.


## Discussion {.challenge}

What other R code from this lesson could we put in a .R file? Or a .Rmd file?

How should we break up a large project (paper/thesis/software package) into files?

What about managing data files?

How should we share a project with others?

* [Software Carpentry's list of best practices in R](http://swcarpentry.github.io/r-novice-inflammation/06-best-practices-R/)


# Packages

Packages are the next step up from sourcing .R files. They let you write code that other people can install and then load with `library`. Packages generally contain documentation for all functions, and one or more vignettes describing how to use the package.

Hadley Wickham has a great package called `devtools` that takes a lot of the pain out of package writing.

```{r, eval=FALSE}
# Create an empty package template
devtools::create("mylibrary")

# ... Edit mylibrary/DESCRIPTION file
# ... Write .R files in mylibrary/R folder

# Load package. Use this during development.
devtools::load_all("mylibrary")

# Build package, including converting inline documentation to .Rd files using roxygen2.
# Check for common problems and missing documentation.
# A CRAN package must pass all checks.
devtools::check("mylibrary")
```

* [R packages book](http://r-pkgs.had.co.nz/)

Packages are most easily distributed by placing them on GitHub. They can then be installed by others using the `devtools` function `install_github`. Once a package is mature and well documented, it can be submitted to CRAN or Bioconductor.

```{r eval=FALSE}
# To install from GitHub:
devtools::install_github("myusername/mylibrary")
```


```{r}
sessionInfo()
```
